### PR TITLE
CHI-1248 use swagger-json for inline swagger definitions

### PR DESCRIPTION
**Pull request by robert-bo-davis**
## Description
This updates swagger-ui-express to eliminate security issues, adds swagger-json, and splits the routes currently covered by swagger into inline comment based schemas/definitions.  

Even splitting the definitions up into comments, swagger definitions are not "lightweight". 

I'm not sure how far I am expected to go with updating all routes. I can continue if this is a priority, but it is fairly time consuming. I'm also going to do another round of double checking that what I have documented is correct. But this gives a general idea of what it looks like. 

I'm going to leave this as a WIP until I get buy in that this is the right direction and clarification around how many endpoints we'd like to cover in this first iteration.

### Related Issues
CHI-1248

### Verification steps
Run your local hrm dev env and check http://localhost:8080/swagger to verify documentation.

### CODE REVIEW

**Comment by stephenhand** [comment]
I think you're right to take a pause for feedback here, would like to hear thoughts from wider team

My 2 cents:

This is certainly a big improvement over everything being off in a separate yaml file. I'm concerned about the size of the annotations required though

I'm of the opinion swagger docs are only delivering marginal value for HRM right now, because the team that maintains HRM is also the team that maintains all the clients - it is an internal back-end rather than a public API in that sense. I thought Swagger was most valuable providing docs to teams other than the maintainers who were implementing clients. Personally I find the service tests and some descriptive inline comments that give some insights into the semantics of the API more useful than swagger docs in our scenario.

But perhaps I'm just making judgements from my own perspectives and other people on the team get more benefit from them than I do. I just think having to have a second representation of a good chunk of the HRM type system is a non trivial amount of maintenance to commit to as a team and I want to be sure the effort is worthwhile.

One thing that makes a case for doing this is that OpenApi does more than Swagger, we can use it for contract tests, for instance. If we had ambitions to leverage Open API for more than we are doing right now, that would definitely up the value of maintaining it.

Another possibility for the future would be code generation. Whilst we might not be able to autogenerate Open API definitions from TS types, there are options for doing it the other way around, have Open API as the source of truth? I'm always a little wary of code generation because it tends to sound great in theory but get a little messy in practice, but it's something to consider.

TL:DR; I'm not convinced the benefit of maintaining Swagger for HRM in either form is worth the ongoing effort unless

a) The rest of the team are getting / planning on getting more use out of the swagger docs than I do (I don't use them) or
b) We have future plans on our roadmap that will either reduce the effort (e.g. code generation) or increase the value (e.g. contract testing) of having OpenAPI - we don't right at the moment but we could,

File: hrm-service/src/case/case-data-access.ts [hunk 1]
```diff
@@ -7,6 +7,37 @@ import { DELETE_BY_ID } from './sql/case-delete-sql';
 import { selectSingleCaseByIdSql } from './sql/case-get-sql';
 import { Contact } from '../contact/contact-data-access';
 
+/**
```
Comment (stephenhand) [inline hrm-service/src/case/case-data-access.ts:10]: We would probably have to refactor our types a little if we were to go down this path - at least move them to other modules

Normally I'm of the opinion that types are best defined as close to the code that uses them as possible, but we definitely don't want HTTP related annotations polluting the data access modules, so we should probably move these types out to their own files in this case

TBF they started life as data access only types but are now used further up the stack, so there's probably no harm moving them out to a separate file anyway

Comment (robert-bo-davis) [reply hrm-service/src/case/case-data-access.ts:10]: 💯

**Review by stephenhand — COMMENTED** [review]
Another questiopn: Do we think there's any value in a halfway solution, where we have OpenAPI definitions, but leave the response and request payloads more loosely defined than they would be in TS, so we can just have some smaller annotations up in the route definitions, rather than having to define lots of components defined for all our types?

Or would that just not be worth doing?

**Comment by robert-bo-davis** [comment]
My thinking was that having the structure definitions right next to the TS definitions that they represent would make it easier and more obvious for devs and therefor more likely that they would keep them up to date.

I totally agree with your assessment that, unless folks are using swagger regularly to do their jobs, the overhead of creating and maintaining the definitions is probably not worth the payoff. It is a fairly time consuming and complex process to perform retroactively. It would probably be less onerous if they were being created when types were being added and as code changes, but I still question the usefulness at this stage. If there is a team of strictly front end devs, that is a different story. If everyone is full stack, I'm not sure if swagger is worth the time and effort.

It was a great first step for me to familiarize myself with HRM either way, though!

**Comment by stephenhand** [comment]
> My thinking was that having the structure definitions right next to the TS definitions that they represent would make it easier and more obvious for devs and therefor more likely that they would keep them up to date.
> 
> I totally agree with your assessment that, unless folks are using swagger regularly to do their jobs, the overhead of creating and maintaining the definitions is probably not worth the payoff. It is a fairly time consuming and complex process to perform retroactively. It would probably be less onerous if they were being created when types were being added and as code changes, but I still question the usefulness at this stage. If there is a team of strictly front end devs, that is a different story. If everyone is full stack, I'm not sure if swagger is worth the time and effort.
> 
> It was a great first step for me to familiarize myself with HRM either way, though!

Yes, - I meant to say, impressed with how much you got done here! You documented a serious chunk of the type system in this PR. It's certainly one way to get up to speed, if not one for the faint of heart 😅

**Review by murilovmachado — COMMENTED** [review]
My 2 cents:
- I think Swagger provides an excellent (live) documentation
- And I think it's very useful (not only for API consumers but anyone)
- But we were not maintaining it, possibly for 2 reasons:
- It takes some (a lot?) effort to maintain it
- Developers forget to do it (this was kinda addressed in this PR moving it closer to the code)

Since Swagger does not take any advantage of the types defined in typescript, I feel like the effort of maintaining it will continue to be an issue, and developers will tend to not keep it up to date.

There're tools like [ts-json-schema-generator](https://github.com/vega/ts-json-schema-generator) which can generate schemas from Typescript, and we could make Swagger be fed from it. I think it would be good to investigate this if we still want to keep Swagger.

There are other tools that do the opposite: generate Typescript from Swagger definitions. Instinctively, I don't like this approach. I would not want a tool generating my TS code.

**Comment by robert-bo-davis** [comment]
I missed a tool in my initial research. There *is* a tool for generating open api definitions from typescript definitions: [tsoa](https://github.com/lukeautry/tsoa)

I'm happy to do a quick PoC of implementing this library instead of manually defining Openapi types to see if that will reduce the dev maintenance overhead significantly.

**Comment by stephenhand** [comment]
> I missed a tool in my initial research. There _is_ a tool for generating open api definitions from typescript definitions: [tsoa](https://github.com/lukeautry/tsoa)
> 
> I'm happy to do a quick PoC of implementing this library instead of manually defining Openapi types to see if that will reduce the dev maintenance overhead significantly.

Yes please! That sounds more along the lines of what we're after, if it works.

I suppose the question we need to answer is what limitations or restrictions does it impose on our type system in order that it can generate a nice OpenAPI definition from them, I have to imagine there will be some.

**Comment by robert-bo-davis** [comment]
After digging in a bit I found that the tsoa decorator system depends on class based controllers. It won't work with our function based controllers. There are a few other potential ways to feed types into json schema including [ts-json-schema-generator](https://github.com/vega/ts-json-schema-generator) and [typebox](https://github.com/sinclairzx81/typebox). 

Based on a shallow dive, these seem like they would simplify keeping type definitions in sync with swagger definitions, but would likely require a bit of work to get set up. I believe that they would also require significant swagger definitions to define endpoints that use the generated types.

I would say that, unless we have developers who plan to actively using the swagger documentation on a daily basis to do their jobs, it is probably not worth the upfront time or maintenance requirements based on my understanding of our current product priorities and team size.

**Comment by nickhurlburt** [comment]
@robert-bo-davis makes sense to me. So where does that bring us to? Are we back to scrapping swagger?

**Comment by stephenhand** [comment]
> @robert-bo-davis makes sense to me. So where does that bring us to? Are we back to scrapping swagger?

I think so - nobody has spoken in up in any of the discussions we've had around this to say that they get a lot of use out of swagger in their development work, so in the absence of any 'cheap' options for autogenerating them, I would say scrapping it and replacing with inline comments would be the best bet.

We can always refer back to this PR for info if we need swagger again, for documenting APIs that will be consumed by 3rd parties for instance ('adapter lambdas' for the push model of transferring resource DB data comes to mind)

**Comment by robert-bo-davis** [comment]
Closing this and working on PR that removes swagger
